src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -865,6 +865,23 @@ public static TypeHandle TypeHandleOf<T>()
         {
             return new TypeHandle((void*)RuntimeTypeHandle.ToIntPtr(typeof(T).TypeHandle));
         }
+
+        public static bool AreSameType(TypeHandle left, TypeHandle right) => left.m_asTAddr == right.m_asTAddr;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool CanCastTo(TypeHandle destTH)
+        {
+            CastResult result = CastCache.TryGet(CastHelpers.s_table!, (nuint)m_asTAddr, (nuint)destTH.m_asTAddr);
+
+            if (result != CastResult.MaybeCast)
+                return result == CastResult.CanCast;
+
+            return CanCastTo_NoCacheLookup(m_asTAddr, destTH.m_asTAddr);
```

Comment:
Nit: The first thing that the QCall is going to do is repeat the cache lookup...

Reply:
Also, the unmanaged `TypeHandle.CanCastTo` assumes that some cases are very cheap to check for and it does not bother to add them to the cache. These cases will be much slower here since we are always take the QCall transition for them. We should either check for them here and/or add them to cache (similar to how `RuntimeTypeHandle::CanCastTo` adds them to the cache for the same reasons).

Reply:
> similar to how `RuntimeTypeHandle::CanCastTo` adds them to the cache for the same reasons

The methods should probably be combined at managed side. They are doing the exact same things.

Reply:
Yes (it is fine to do it in a follow up PR).

Reply:
> They are doing the exact same things.

In fact they aren't. `RuntimeTypeHandle::CanCastTo` allows `T -> Nullable<T>`.

The non-cached cases include nullables, COM and I(Dynamic)Castable interfaces. I don't think specially handling them is worthy, even for future callers.

Reply:
Yes, you would either need to have a bool flag that controls the special handling or introduce two QCalls with similar implementation.

Reply:
Do you plan to do something about this one?

Reply:
Yes, together with some other cases sharable between `RuntimeTypeHandle` and `MethodTable`, like type equivalence.